editor/utils/dom.js

```diff
@@ -56,18 +57,85 @@ export function isEdge( container, start = false ) {
 }
 
 /**
- * Places the caret at start or end of a given element
+ * Holds the first caret boundary rectangle for subsequent calls.
+ *
+ * @type {DOMRect}
+ */
+let firstVerticalRect;
+
+/**
+ * Resets the first caret position.
+ */
+export function resetVerticalPosition() {
+	firstVerticalRect = null;
+}
+
+/**
+ * Check whether the caret is vertically at the edge of the container.
  *
- * @param  {Element} container DOM Element
- * @param  {Boolean} start     Position: Start or end of the element
+ * @param  {Element} container Focusable element.
+ * @param  {Boolean} reverse   Set to true to check top, false for bottom.
+ * @return {Boolean}           True if at the edge, false if not.
  */
-export function placeCaretAtEdge( container, start = false ) {
+export function isVerticalEdge( container, reverse ) {
+	if ( [ 'INPUT', 'TEXTAREA' ].indexOf( container.tagName ) !== -1 ) {
```

Comment:
Minor: naming this, as is done in other parts of the diff, would be nice: `const isInputOrTextarea =`

Reply:
I almost changed the other one. :) What makes you prefer it?

Reply:
That `indexOf` isn't very descriptive and the variable assignment counters that. I would also go, instead, for `lodash#includes`.